src/jsonapi.js

```diff
@@ -265,113 +265,114 @@ export const requireEntity = (entityType, endpoint = entityType, {
+    return imm(state).set(['endpoint', 'path'], path).value();
   },
 
   [API_WILL_CREATE]: (state) => {
-    return Imm.fromJS(state).update('isCreating', v => v + 1).toJS();
+    return imm(state).set(['isCreating'], state.isCreating + 1).value();
   },
 
-  [API_CREATED]: (rawState, { payload: rawEntities }) => {
-    const state = Imm.fromJS(rawState);
-    const entities = Imm.fromJS(
-      Array.isArray(rawEntities) ? rawEntities : [rawEntities]
+  [API_CREATED]: (state, { payload: resources }) => {
+    const newState = updateOrInsertResourcesIntoState(
+      state,
+      Array.isArray(resources) ? resources : [resources]
     );
 
-    return updateOrInsertEntitiesIntoState(state, entities)
-      .update('isCreating', v => v - 1)
-      .toJS();
+    return imm(newState)
+      .set('isCreating', state.isCreating - 1)
+      .value();
   },
 
   [API_CREATE_FAILED]: (state) => {
-    return Imm.fromJS(state).update('isCreating', v => v - 1).toJS();
+    return imm(state).set(['isCreating'], state.isCreating - 1).value();
   },
 
   [API_WILL_READ]: (state) => {
-    return Imm.fromJS(state).update('isReading', v => v + 1).toJS();
+    return imm(state).set(['isReading'], state.isReading + 1).value();
   },
 
-  [API_READ]: (rawState, { payload }) => {
-    const state = Imm.fromJS(rawState);
-    const entities = Imm.fromJS(
-      Array.isArray(payload.data) ? payload.data : [payload.data]
-    ).concat(Imm.fromJS(payload.included || []));
+  [API_READ]: (state, { payload }) => {
+    const resources = (
+      Array.isArray(payload.data)
+        ? payload.data
+        : [payload.data]
+    ).concat(payload.included || []);
 
-    return updateOrInsertEntitiesIntoState(state, entities)
-      .update('isReading', v => v - 1)
-      .toJS();
+    const newState = updateOrInsertResourcesIntoState(state, resources);
+
+    return imm(newState)
+      .set('isReading', state.isReading - 1)
+      .value();
   },
 
   [API_READ_FAILED]: (state) => {
-    return Imm.fromJS(state).update('isReading', v => v - 1).toJS();
+    return imm(state).set(['isReading'], state.isReading - 1).value();
   },
 
-  [API_WILL_UPDATE]: (rawState, { payload: entity }) => {
-    const { type, id } = entity;
-    const state = Imm.fromJS(rawState);
+  [API_WILL_UPDATE]: (state, { payload: resource }) => {
+    const { type, id } = resource;
 
-    return setIsInvalidatingForExistingEntity(state, { type, id }, IS_UPDATING)
-      .update('isUpdating', v => v + 1)
-      .toJS();
+    return setIsInvalidatingForExistingResource(state, { type, id }, IS_UPDATING)
+      .set('isUpdating', state.isUpdating + 1)
+      .value();
   },
 
-  [API_UPDATED]: (rawState, { payload: rawEntities }) => {
-    const state = Imm.fromJS(rawState);
-    const entities = Imm.fromJS(
-      Array.isArray(rawEntities) ? rawEntities : [rawEntities]
+  [API_UPDATED]: (state, { payload: resources }) => {
+    const newState = updateOrInsertResourcesIntoState(
+      newState,
```

Comment:
Possible
`const newState = updateOrInsertResourcesIntoState(newState, Array.isArray(resources) ? resources : [resources])` ??? shouldn't the aregument newState be state ?